### ISSUE TITLE
Unable to make interfaces group by using the same funtion interface implenentations

### ISSUE THREAD

**Issue by erickskrauch** [goava/di#11]
The title :)
https://play.golang.org/p/ZvWhZaxyzM6
```go
package main

import (
	"context"
	"log"

	"github.com/goava/di"
)

type Interface interface {
	Do(ctx context.Context) error
}

type FuncInterface func(ctx context.Context) error
func (c FuncInterface) Do(ctx context.Context) error {
	return c(ctx)
}

func createFirst() FuncInterface {
	return func(ctx context.Context) error {
		return nil
	}
}

func createSecond() FuncInterface {
	return func(ctx context.Context) error {
		return nil
	}
}

func main() {
	container, err := di.New(
		di.Provide(createFirst, di.As(new(Interface))),
		di.Provide(createSecond, di.As(new(Interface))),
	)
	if err != nil {
		log.Fatal(err)
	}

	var interfaces []Interface
	err = container.Resolve(&interfaces)
	if err != nil {
		log.Fatal(err)
	}
}
```

The error:
```
main.FuncInterface already exists in dependency graph
```

**Comment by defval** [comment]
Yes, that would be convenient. I will add this usage case. Thank you very much for your reports.

**Comment by erickskrauch** [comment]
I always use libraries on the edge of their capabilities :smile:

**Comment by erickskrauch** [comment]
@defval, excuse me for asking, but is there any ETA to implement this case? Without this feature, I can't continue implementing the DI.

If you are busy, I can try to prepare the necessary changes myself and send PR if I succeed :)

**Comment by defval** [comment]
@erickskrauch You can use `di.WithName()` as a temporary solution:

```go
di.Provide(createFirst, di.As(new(Interface)), di.WithName("1")),
di.Provide(createSecond, di.As(new(Interface)), di.WithName("2")),
```

Currently, the type store is designed so that there can only be one unique constructor for the type. Because of it `map[typeKey]provider`. Refactoring to the `map[typeKey][]provider` will lead to a large number of changes.

I need some time to think about it :)

**Comment by defval** [comment]
I would like to prepare and release the existing features. And then start this one.

**Comment by erickskrauch** [comment]
Thanks for the workaround!

Is it also possible to ensure that the original constructor's type isn't lost when `di.WithName()` and `di.As()` are applied to it?

**Comment by defval** [comment]
> Is it also possible to ensure that the original constructor's type isn't lost when di.WithName() and di.As() are applied to it?

Sure. 

 - `di.As()` adds an alias that points to the original constructor. 
 - `di.WithName()` does not modify the constructor in any way. It modifies constructor identity.

**Comment by erickskrauch** [comment]
```go
package main

import (
	"log"

	"github.com/goava/di"
)

type Interface interface {
	Do()
}

type Implementation struct {}
func (d Implementation) Do() {}

func newImplementation() *Implementation {
	return &Implementation{}
}

func main() {
	container, err := di.New(
		di.Provide(newImplementation, di.WithName("hello")),
	)
	if err != nil {
		log.Fatal(err)
	}

	var implementation *Implementation
	err = container.Resolve(&implementation)
	if err != nil {
		log.Fatal(err)
	}
}
```

Output:
```
2020/04/18 22:20:43 main.go:35: type *main.Implementation not exists in container
```

**Comment by defval** [comment]
If you use `di.WithName()` in provide, you need to use `di.Name()` with resolve.

```go
container.Resolve(&implementation, di.Name("hello"))
```

**Comment by erickskrauch** [comment]
Yes. But I think it's better to create a named alias than to replace the type. Perhaps you should introduce an error when trying to create several constructors that create a structure without giving it a name. In other words, you should let it work without an error as shown in the code sample above but throw an error when trying to create a second constructor that will create the same structure without giving it a name.

**Comment by erickskrauch** [comment]
Resolved by #14.
